app/controllers/api/storage.php

```diff
@@ -306,97 +307,200 @@
-        
-        if (!$device->upload($file['tmp_name'], $path)) { // TODO deprecate 'upload' and replace with 'move'
-            throw new Exception('Failed moving file', 500);
-        }
+        $size = $size ?? $device->getFileSize($fileTmpName);
+        $path = $device->getPath($fileId . '.' . \pathinfo($fileName, PATHINFO_EXTENSION));
+        $path = str_ireplace($device->getRoot(), $device->getRoot() . DIRECTORY_SEPARATOR . $bucket->getId(), $path);
 
-        $mimeType = $device->getFileMimeType($path); // Get mime-type before compression and encryption
-
-        if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
-            $antiVirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
-                (int) App::getEnv('_APP_STORAGE_ANTIVIRUS_PORT', 3310));
+        $file = $dbForInternal->getDocument('files', $fileId);
 
-            if (!$antiVirus->fileScan($path)) {
-                $device->delete($path);
-                throw new Exception('Invalid file', 403);
+        if (!$file->isEmpty()) {
+            $chunks = $file->getAttribute('totalChunks', 1);
+            if ($chunk == -1) {
+                $chunk = $chunks - 1;
             }
         }
 
-        // Compression
-        $data = $device->read($path);
-        if($size <= APP_LIMIT_COMPRESSION) {
-            $compressor = new GZIP();
-            $data = $compressor->compress($data);
-        }
-        
-        if($bucket->getAttribute('encryption', true) && $size <= APP_LIMIT_ENCRYPTION) {
-            $key = App::getEnv('_APP_OPENSSL_KEY_V1');
-            $iv = OpenSSL::randomPseudoBytes(OpenSSL::cipherIVLength(OpenSSL::CIPHER_AES_128_GCM));
-            $data = OpenSSL::encrypt($data, OpenSSL::CIPHER_AES_128_GCM, $key, 0, $iv, $tag);
+        $chunksUploaded = $device->upload($fileTmpName, $path, $chunk, $chunks);
+        if (empty($chunksUploaded)) {
+            throw new Exception('Failed uploading file', 500);
         }
 
-        if (!$device->write($path, $data, $mimeType)) {
-            throw new Exception('Failed to save file', 500);
-        }
+        if ($chunksUploaded == $chunks) {
+            $mimeType = $device->getFileMimeType($path); // Get mime-type before compression and encryption
+
+            if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
+                $antiVirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
+                    (int) App::getEnv('_APP_STORAGE_ANTIVIRUS_PORT', 3310));
+
+                if (!$antiVirus->fileScan($path)) {
+                    $device->delete($path);
+                    throw new Exception('Invalid file', 403);
+                }
+            }
+
+            // Compression
+            if($size <= APP_LIMIT_COMPRESSION || $size <= APP_LIMIT_ENCRYPTION) {
```

Comment:
If one of this conditions will be true both actions will happen. We should probably decouple them to separate blocks

Reply:
yeh, I just forgot to push the fix.